In [1]:
import numpy as np
import pandas as pd
import datetime

In [2]:
p1_gen = pd.read_csv("Plant_1/Plant_1_Generation_Data.csv")
p1_weath = pd.read_csv("Plant_1/Plant_1_Weather_Sensor_Data.csv")

p2_gen = pd.read_csv("Plant_2/Plant_2_Generation_Data.csv")
p2_weath = pd.read_csv("Plant_2/Plant_2_Weather_Sensor_Data.csv")

In [3]:
# CONTROL - all dates are present
dff = p1_gen
dff['DATE'] = pd.to_datetime(dff['DATE_TIME'], format="%d-%m-%Y %H:%M").dt.date
print(dff['DATE'].unique())

[datetime.date(2020, 5, 15) datetime.date(2020, 5, 16)
 datetime.date(2020, 5, 17) datetime.date(2020, 5, 18)
 datetime.date(2020, 5, 19) datetime.date(2020, 5, 20)
 datetime.date(2020, 5, 21) datetime.date(2020, 5, 22)
 datetime.date(2020, 5, 23) datetime.date(2020, 5, 24)
 datetime.date(2020, 5, 25) datetime.date(2020, 5, 26)
 datetime.date(2020, 5, 27) datetime.date(2020, 5, 28)
 datetime.date(2020, 5, 29) datetime.date(2020, 5, 30)
 datetime.date(2020, 5, 31) datetime.date(2020, 6, 1)
 datetime.date(2020, 6, 2) datetime.date(2020, 6, 3)
 datetime.date(2020, 6, 4) datetime.date(2020, 6, 5)
 datetime.date(2020, 6, 6) datetime.date(2020, 6, 7)
 datetime.date(2020, 6, 8) datetime.date(2020, 6, 9)
 datetime.date(2020, 6, 10) datetime.date(2020, 6, 11)
 datetime.date(2020, 6, 12) datetime.date(2020, 6, 13)
 datetime.date(2020, 6, 14) datetime.date(2020, 6, 15)
 datetime.date(2020, 6, 16) datetime.date(2020, 6, 17)]


In [4]:
#dff[ (dff['DATE_TIME'].dt.date == datetime.date(2020, 10,6) )]

In [5]:
for df in [p1_gen]:
    df['DATE_TIME'] = pd.to_datetime(df['DATE_TIME'], format="%d-%m-%Y %H:%M")
    
for df in [p1_weath, p2_gen, p2_weath]:
    df['DATE_TIME'] = pd.to_datetime(df['DATE_TIME'], format="%Y-%m-%d %H:%M:%S")

In [6]:
df_1g = p1_gen.copy()
df_1w = p1_weath.copy()
df_2g = p2_gen.copy()
df_2w = p2_weath.copy()

#for df in [df_1g, df_1w, df_2g, df_2w]:
df_1g = df_1g[ (df_1g['DATE_TIME'].dt.time >= datetime.time(6, 0) ) & (df_1g['DATE_TIME'].dt.time <= datetime.time(18, 0)) ]
df_1w = df_1w[ (df_1w['DATE_TIME'].dt.time >= datetime.time(6, 0) ) & (df_1w['DATE_TIME'].dt.time <= datetime.time(18, 0)) ]

df_2g = df_2g[ (df_2g['DATE_TIME'].dt.time >= datetime.time(6, 0) ) & (df_2g['DATE_TIME'].dt.time <= datetime.time(18, 0)) ]
df_2w = df_2w[ (df_2w['DATE_TIME'].dt.time >= datetime.time(6, 0) ) & (df_2w['DATE_TIME'].dt.time <= datetime.time(18, 0)) ]

In [7]:
df_1g

DATE_TIME  PLANT_ID       SOURCE_KEY    DC_POWER   AC_POWER  \
510   2020-05-15 06:00:00   4135001  1BY6WEcLGh8j5v7   37.142857   3.585714   
511   2020-05-15 06:00:00   4135001  1IF53ai7Xc0U56Y   53.500000   5.162500   
512   2020-05-15 06:00:00   4135001  3PZuoBAID5Wc2HD   58.000000   5.585714   
513   2020-05-15 06:00:00   4135001  7JYdWkrLSPkdwr4   58.428571   5.628571   
514   2020-05-15 06:00:00   4135001  McdE0feGgRqW7Ca   54.375000   5.250000   
...                   ...       ...              ...         ...        ...   
68267 2020-06-17 18:00:00   4135001  uHbuxQJl8lW7ozc  615.857143  59.614286   
68268 2020-06-17 18:00:00   4135001  wCURE6d3bPkepu2  600.714286  58.142857   
68269 2020-06-17 18:00:00   4135001  z9Y9gH1T5YWrNuG  597.857143  57.885714   
68270 2020-06-17 18:00:00   4135001  zBIq5rxdHJRwDNY  594.000000  57.528571   
68271 2020-06-17 18:00:00   4135001  zVJPv84UY57bAof  603.000000  58.414286   

       DAILY_YIELD  TOTAL_YIELD        DATE  
510       0.000000  6259559.000  2020-05-15  
511       0.000000  6183645.000  2020-05-15  
512       0.000000  6987759.000  2020-05-15  
513       0.000000  7602960.000  2020-05-15  
514       0.000000  7158964.000  2020-05-15  
...            ...          ...         ...  
68267  5949.000000  7286984.000  2020-06-17  
68268  5866.571429  7028584.571  2020-06-17  
68269  5802.000000  7251187.000  2020-06-17  
68270  5799.000000  6583351.000  2020-06-17  
68271  5892.000000  7363254.000  2020-06-17  

[35888 rows x 8 columns]

# ON ALL DATA (including night-time)

In [8]:
# p1_gen, p1_weath, p2_gen, p2_weath

dc = []

gen_zeros = []
for df in [p1_gen, p2_gen]:
    DC = (df['DC_POWER'] == 0).sum()
    AC = (df['AC_POWER'] == 0).sum()
    DY = (df['DAILY_YIELD'] == 0).sum()
    TY = (df['TOTAL_YIELD'] == 0).sum()
    zeros = DC + AC + DY + TY
    gen_zeros.append(zeros)
    dc.append(DC)
    
weath_zeros = []
for df in [p1_weath, p2_weath]:
    amb = (df['AMBIENT_TEMPERATURE'] == 0).sum()
    mod = (df['MODULE_TEMPERATURE'] == 0).sum()
    irr = (df['IRRADIATION'] == 0).sum()
    zeros = amb + mod + irr
    weath_zeros.append(zeros)

In [9]:
print(dc)
print(dc[0]/len(p1_gen))
print(dc[1]/len(p2_gen))

print(gen_zeros)
print(weath_zeros)
print((gen_zeros[0]+weath_zeros[0])/(len(p1_gen)*4+len(p1_weath)*3))
print((gen_zeros[1]+weath_zeros[1])/(len(p2_gen)*4+len(p2_weath)*3))

[31951, 35662]
0.46455261856989155
0.5267807025318325
[82598, 83456]
[1425, 1397]
0.29517174995960066
0.3024318438601556


# ON ONLY DAYTIME DATA

In [10]:
dc = []

gen_zeros = []
for df in [df_1g, df_2g]:
    DC = (df['DC_POWER'] == 0).sum()
    AC = (df['AC_POWER'] == 0).sum()
    DY = (df['DAILY_YIELD'] == 0).sum()
    TY = (df['TOTAL_YIELD'] == 0).sum()
    zeros = DC + AC + DY + TY
    gen_zeros.append(zeros)
    dc.append(DC)
    
weath_zeros = []
for df in [df_1w, df_2w]:
    amb = (df['AMBIENT_TEMPERATURE'] == 0).sum()
    mod = (df['MODULE_TEMPERATURE'] == 0).sum()
    irr = (df['IRRADIATION'] == 0).sum()
    zeros = amb + mod + irr
    weath_zeros.append(zeros)

In [11]:
print(dc)
print(dc[0]/len(df_1g))
print(dc[1]/len(df_2g))

print(gen_zeros)
print(weath_zeros)
print((gen_zeros[0]+weath_zeros[0])/(len(df_1g)*4+len(df_1w)*3))
print((gen_zeros[1]+weath_zeros[1])/(len(df_2g)*4+len(df_2w)*3))

[217, 4003]
0.006046589389210879
0.1160088100620182
[709, 9154]
[0, 1]
0.004775793664158645
0.06401650234249354


##### Plant 1 zeros
Gen = 709, all = 143 552

Weather = 0, all = 4905

Together = 709/148 457 = 0.005 -> 0.5%

DC = 217/35 888 = 0.006 -> 0.6% (217/709 -> 30% from all missing values)

##### Plant 2 zeros
Gen = 9154, all = 138 024

Weather = 1, all = 4986

Together = 9155/143 010 = 0.064 -> 6.4%

DC = 4003/34 506 = 0.116 -> 11.6 or 12% (4003/9154 -> 44% of all missing values)